cluster/src/main/java/io/scalecube/cluster/Cluster.java

```diff
@@ -141,19 +100,37 @@ public static ICluster joinAwait(ClusterConfig config) {
   }
 
   public static ListenableFuture<ICluster> join(final ClusterConfig config) {
-    if (config.portAutoIncrement) { // Find available port
-      config.port = AvailablePortFinder.getNextAvailable(config.port, config.portCount);
-    }
     return new Cluster(config).join0();
   }
 
   private ListenableFuture<ICluster> join0() {
-    updateClusterState(State.INSTANTIATED, State.JOINING);
     LOGGER.info("Cluster instance '{}' joining seed members: {}", memberId, config.seedMembers);
-    ListenableFuture<Void> transportFuture = transport.start();
-    ListenableFuture<Void> clusterFuture = transform(transportFuture, new AsyncFunction<Void, Void>() {
+    ListenableFuture<Transport> transportFuture = Transport.bind(config.transportConfig);
+    ListenableFuture<Void> clusterFuture = transform(transportFuture, new AsyncFunction<Transport, Void>() {
       @Override
-      public ListenableFuture<Void> apply(@Nullable Void param) throws Exception {
+      public ListenableFuture<Void> apply(@Nullable Transport input) throws Exception {
+        // Init transport component
+        transport = input;
+
+        // Init gossip protocol component
+        gossipProtocol = new GossipProtocol(memberId, transport, config.gossipProtocolConfig);
+
+        // Init failure detector component
+        failureDetector = new FailureDetector(transport, config.failureDetectorConfig);
+
+        // Init cluster membership component
+        clusterMembership = new ClusterMembership(memberId, transport);
+        clusterMembership.setFailureDetector(failureDetector);
+        clusterMembership.setGossipProtocol(gossipProtocol);
+        clusterMembership.setLocalMetadata(config.metadata);
+        clusterMembership.setSeedMembers(config.seedMembers);
```

Comment:
Next setters better to be removed :   ``setSyncTime``, ``setSyncTimeout``, ``setSyncGroup``, ``setMaxSuspectTime``, ``setMaxShutdownTime`` . Use ``config`` instead .